### ISSUE TITLE
False positive when @template tag refers to a template param not defined before the current line

### ISSUE THREAD

**Issue by ste93cry** [vimeo/psalm#4293]
The following code is valid, however when PSalm analyzes it it thinks that the `TResult` type of the `QueryHandlerInterface` interface is not defined. Exchanging the order of the two `@template` tags solves the issue. As a user I would expect that it works both ways, since the only case where the analyzer should complain is if `TResult` is really undefined

```php
/**
 * @template TQuery of QueryInterface<TResult>
 * @template TResult
 */
interface QueryHandlerInterface
{
    /**
     * @param TQuery $query
     *
     * @return TResult
     */
    public function handle(QueryInterface $query);
}

/**
 * @template TResult
 */
interface QueryInterface
{
}

/**
 * @template-implements QueryInterface<FooQueryResult>
 */
final class FooQuery implements QueryInterface
{
}

final class FooQueryResult
{
}

/**
 * @template-implements QueryHandlerInterface<FooQuery, FooQueryResult>
 */
final class FooQueryHandler implements QueryHandlerInterface
{
    public function handle(QueryInterface $query): FooQueryResult
    {
        return new FooQueryResult();
    }
}
```

The thrown error is:

```
ERROR: UndefinedDocblockClass - 12:15 - Docblock-defined class or interface TResult does not exist

ERROR: InvalidTemplateParam - 40:40 - Extended template param TQuery expects type QueryInterface<TResult>, type FooQuery given
```

https://psalm.dev/r/8e6f50d591

**Comment by muglug** [comment]
You need to swap the param order for inference to work: https://psalm.dev/r/938d1e73cb

**Comment by ste93cry** [comment]
>You need to swap the param order for inference to work

Indeed, and this is exactly the issue. It should not matter in which order they are, once they are defined they should work. I don't know why you closed the issue, but to me it looks like a bug. Switching the order may solve the issue, sometimes having one parameter before the other makes more sense

**Comment by ste93cry** [comment]
I tracked down the issue to the following code:

https://github.com/vimeo/psalm/blob/7195275993eb7f74a14499b9d8e91a5d09aaa117/src/Psalm/Internal/PhpVisitor/ReflectorVisitor.php#L1190-L1195

Since the sorting affects the subsequent loop statement, when the `TypeParser::parseTokens()` method is called `$storage->template_types` does not contains yet `TResult` which produces a `TNamedObject` template type instead of a `TTemplateParam`. If there was some way to sort `$docblock_info->templates` taking into consideration nested types and the dependencies between tags rather than only the offset appears then everything would work fine (I tested it by reversing the condition of the compare function). @muglug Do you think that this would be possible to do?

**Comment by muglug** [comment]
I'm very happy to accept a PR!

I see that TypeScript does not complain about a similar order, but [TypeScript also fails to catch bugs](https://www.typescriptlang.org/play?#code/FASwdgLgpgTgZgQwMZQAQEUCusCeAJBMAEwBtYBJSWRFAHgBUtdUoAPaYgZw2xh0ujxkUBgCUonTCQgA+ADSp64ydJnAA3sFTbUAC0KkoACgCOvHAC5FTPgEpUVpRKkQA3MAC+wUFSEoeuALUwmLOqhqe3kgkCJzcAGIA9ok2OKggALYADmQZUJDcqUF+Ikkp5souappewNGxCcmpldIRtfVxqABCCDDNYRBtUTGdZakExGQw6dm5+RCF5hOGMMU0pU3mCj19FQPVWjr6k8ZmuFZj5rYXm7gtgzqomo+PMFAQmDBgqGBQAO6oS53AZGWzuR5eDxAA) due to incorrect params being sent.
